### ISSUE TITLE
Spring Mongo JPA - the basics

### ISSUE THREAD

**Issue by Pasquale-Favella** [Pasquale-Favella/pasquale-favella.github.io#9]
Mongo JPA, or the Java Persistence API, is a powerful tool that allows developers to interact with MongoDB databases using the standard JPA interface. JPA is a specification for object-relational mapping in Java, which means that it allows developers to map Java classes to database tables and perform CRUD (Create, Read, Update, Delete) operations on those tables using standard Java methods.

MongoDB is a popular NoSQL database that is widely used in modern web applications due to its flexibility and scalability. With the rise of microservices architecture, many developers are turning to MongoDB as a preferred database solution. However, working with MongoDB can be challenging for developers who are not familiar with NoSQL concepts and query languages.

That's where Mongo JPA comes in. It provides a simple and intuitive way to work with MongoDB databases, allowing developers to focus on their business logic rather than database implementation details. By using Mongo JPA, developers can take advantage of the powerful features of MongoDB, such as schemaless data modeling, dynamic queries, and horizontal scaling.

One of the key benefits of using Mongo JPA is that it provides a consistent API for working with different types of databases, including relational databases and NoSQL databases. This means that developers can use the same Java code to interact with MongoDB and other databases, which can simplify their codebase and reduce the learning curve for new developers.

Another advantage of using Mongo JPA is that it supports the JPA Criteria API, which allows developers to create type-safe queries using Java code. This can help prevent errors and make code easier to maintain, especially for larger projects. Additionally, Mongo JPA supports advanced query features such as aggregation, sorting, and pagination, which can be challenging to implement using raw MongoDB queries.

To use Mongo JPA, developers can add the necessary dependencies to their project and create entity classes that map to MongoDB collections. They can then use standard JPA methods to interact with the database, such as `find` , `persist` , `merge` , and `remove`. Mongo JPA also provides annotations that can be used to define indexes, map embedded objects, and specify other database-related details.
Here an example
``` java
@Entity
public class User {
    
    @Id
    private String id;
    
    @Column(name = "username")
    private String username;
    
    @Column(name = "password")
    private String password;
    
    @Embedded
    private Address address;
    
    // Getters and setters
    
    public String getId() {
        return id;
    }
    
    public void setId(String id) {
        this.id = id;
    }
    
    public String getUsername() {
        return username;
    }
    
    public void setUsername(String username) {
        this.username = username;
    }
    
    public String getPassword() {
        return password;
    }
    
    public void setPassword(String password) {
        this.password = password;
    }
    
    public Address getAddress() {
        return address;
    }
    
    public void setAddress(Address address) {
        this.address = address;
    }
    
    // Embedded class
    
    @Embeddable
    public static class Address {
        
        @Column(name = "street")
        private String street;
        
        @Column(name = "city")
        private String city;
        
        @Column(name = "state")
        private String state;
        
        @Column(name = "zip")
        private String zip;
        
        // Getters and setters
        
        public String getStreet() {
            return street;
        }
        
        public void setStreet(String street) {
            this.street = street;
        }
        
        public String getCity() {
            return city;
        }
        
        public void setCity(String city) {
            this.city = city;
        }
        
        public String getState() {
            return state;
        }
        
        public void setState(String state) {
            this.state = state;
        }
        
        public String getZip() {
            return zip;
        }
        
        public void setZip(String zip) {
            this.zip = zip;
        }
    }
}
```
In this example, the `User` class is annotated with `@Entity`, which tells Mongo JPA that it should be mapped to a MongoDB collection. The `@Id` annotation is used to indicate that the `id` field is the primary key for the collection. The `@Column` annotation is used to specify the names of the fields in the collection.
The `Address` class is embedded in the `User` class using the `@Embedded` annotation. It is also annotated with `@Embeddable`, which indicates that it should be stored as part of the `User` document in the MongoDB collection.

The class above can be used to build a `MongoRepository`
``` java
@Repository
public interface UserRepository extends MongoRepository<User, String> {
    
    User findByUsername(String username);
    
    List<User> findByAddressCity(String city);
    
    @Query("{ 'address.state' : ?0 }")
    List<User> findByAddressState(String state);
    
    List<User> findByAddressStateAndAddressCity(String state, String city);
    
    default List<User> findUsersByCriteria(String state, String city) {
        Criteria criteria = new Criteria();
        
        if (state != null && !state.isEmpty()) {
            criteria = criteria.and("address.state").is(state);
        }
        
        if (city != null && !city.isEmpty()) {
            criteria = criteria.and("address.city").is(city);
        }
        
        Query query = new Query(criteria);
        return mongoTemplate.find(query, User.class);
    }
}
```
In this example, the `findUsersByCriteria` method is a custom query that uses `Criteria` to build a dynamic query. `Criteria` is a flexible way to build complex queries with conditions that can change at runtime.
The method takes two parameters, `state` and `city`, which are used to build the `Criteria` object. The method first creates a new `Criteria` object and then adds conditions based on the values of `state` and `city`. If the `state` parameter is not null and not empty, it adds a condition to match the `state` field in the `address` object. If the `city` parameter is not null and not empty, it adds a condition to match the `city` field in the `address` object.
Finally, the method creates a new `Query` object with the `Criteria` object and uses `MongoTemplate` to execute the query and return a list of `User` objects.

In conclusion, Mongo JPA is a valuable tool for developers who want to work with MongoDB databases using standard Java methods. It provides a consistent and intuitive API that can simplify code and reduce the learning curve for new developers. By leveraging the powerful features of MongoDB, developers can build scalable and flexible applications that can meet the demands of modern web development.
